False-positive `F821: Undefined name` with `mypy_extensions.DefaultNamedArg`

```python
from typing import Any, Callable

from mypy_extensions import DefaultNamedArg

SomeCallableT = Callable[[Any, DefaultNamedArg(bool | None, name="some_prop_name")], None]
```

```console
$ ruff test_ruff_F821.py
Found 1 error(s).
test_ruff_F821.py:5:66: F821 Undefined name `some_prop_name`
```

--- a/resources/test/fixtures/pyflakes/F821_7.py
+++ b/resources/test/fixtures/pyflakes/F821_7.py
@@ -1,0 +1,13 @@
+"""Test: Mypy extensions."""
+
+from mypy_extensions import DefaultNamedArg
+
+# OK
+_ = DefaultNamedArg(bool | None, name="some_prop_name")
+_ = DefaultNamedArg(type=bool | None, name="some_prop_name")
+_ = DefaultNamedArg(bool | None, "some_prop_name")
+
+# Not OK
+_ = DefaultNamedArg("Undefined", name="some_prop_name")
+_ = DefaultNamedArg(type="Undefined", name="some_prop_name")
+_ = DefaultNamedArg("Undefined", "some_prop_name")

--- a/src/checkers/ast.rs
+++ b/src/checkers/ast.rs
@@ -2507,6 +2507,29 @@ where
                         self.visit_expr(value);
                         self.in_type_definition = prev_in_type_definition;
                     }
+                } else if ["Arg", "DefaultArg", "NamedArg", "DefaultNamedArg"]
+                    .iter()
+                    .any(|target| {
+                        match_call_path(&call_path, "mypy_extensions", target, &self.from_imports)
+                    })
+                {
+                    self.visit_expr(func);
+
+                    // Ex) DefaultNamedArg(bool | None, name="some_prop_name")
+                    let mut arguments = args.iter().chain(keywords.iter().map(|keyword| {
+                        let KeywordData { value, .. } = &keyword.node;
+                        value
+                    }));
+                    if let Some(expr) = arguments.next() {
+                        self.in_type_definition = true;
+                        self.visit_expr(expr);
+                        self.in_type_definition = prev_in_type_definition;
+                    }
+                    for expr in arguments {
+                        self.in_type_definition = false;
+                        self.visit_expr(expr);
+                        self.in_type_definition = prev_in_type_definition;
+                    }
                 } else {
                     visitor::walk_expr(self, expr);
                 }

--- a/src/pyflakes/mod.rs
+++ b/src/pyflakes/mod.rs
@@ -96,6 +96,7 @@ mod tests {
     #[test_case(CheckCode::F821, Path::new("F821_4.py"); "F821_4")]
     #[test_case(CheckCode::F821, Path::new("F821_5.py"); "F821_5")]
     #[test_case(CheckCode::F821, Path::new("F821_6.py"); "F821_6")]
+    #[test_case(CheckCode::F821, Path::new("F821_7.py"); "F821_7")]
     #[test_case(CheckCode::F822, Path::new("F822.py"); "F822")]
     #[test_case(CheckCode::F823, Path::new("F823.py"); "F823")]
     #[test_case(CheckCode::F831, Path::new("F831.py"); "F831")]

